### DISCUSSION TITLE
Build fails. Dependencies missing?

### DISCUSSION THREAD

**Discussion by muchgitsuchwow** [YaLTeR/niri#179] [Ideas]
Recently I cannot build niri any longer (using a slightly modified version of the Arch AUR PKGBUILD):

```
   Compiling niri v0.1.1 (/home/user/.local/var/lib/makepkg/aur/niri-git/src/niri)
error: linking with `cc` failed: exit status: 1
  |
  = note: LC_ALL="C" PATH="/usr/lib64/rustlib/x86_64-unknown-linux-gnu/bin:/home/user/.local/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/sbin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl" VSLANG="1033" "cc" "-m64" "/tmp/rustcFzsUOh/symbols.o" "/home/user/.local/var/lib/makepkg/aur/niri-git/src/niri/target/x86_64-unknown-linux-gnu/aur/deps/niri-af62f87027fe0571.niri.1be11338d174d065-cgu.0.rcgu.o" "-Wl,--as-needed" "-L" "/home/user/.local/var/lib/makepkg/aur/niri-git/src/niri/target/x86_64-unknown-linux-gnu/aur/deps" "-L" "/home/user/.local/var/lib/makepkg/aur/niri-git/src/niri/target/aur/deps" "-L" "/usr/lib" "-L" "/usr/lib" "-L" "/usr/lib" "-L" "/usr/lib" "-L" "/home/user/.local/var/lib/makepkg/aur/niri-git/src/niri/target/x86_64-unknown-linux-gnu/aur/build/libspa-dcedcb7111ff4692/out" "-L" "/home/user/.local/var/lib/makepkg/aur/niri-git/src/niri/target/x86_64-unknown-linux-gnu/aur/build/libspa-sys-e6e2b4aafdbc1817/out" "-L" "/usr/lib64/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "/tmp/rustcFzsUOh/liblibspa-75b56344d3d3baef.rlib" "/tmp/rustcFzsUOh/liblibspa_sys-4b781d0f87d8f4ec.rlib" "/usr/lib64/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-8107a46244b29f2b.rlib" "-Wl,-Bdynamic" "-lpipewire-0.3" "-lpipewire-0.3" "-lpangocairo-1.0" "-lpango-1.0" "-lgobject-2.0" "-lglib-2.0" "-lharfbuzz" "-lcairo" "-lpango-1.0" "-lgobject-2.0" "-lglib-2.0" "-lharfbuzz" "-lgobject-2.0" "-lgio-2.0" "-lgobject-2.0" "-lglib-2.0" "-lgobject-2.0" "-lglib-2.0" "-lcairo-gobject" "-lcairo" "-lgobject-2.0" "-lglib-2.0" "-lgobject-2.0" "-lglib-2.0" "-lgbm" "-lxkbcommon" "-lseat" "-ludev" "-linput" "-ldl" "-lpixman-1" "-ldl" "-lgbm" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/usr/lib64/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/home/user/.local/var/lib/makepkg/aur/niri-git/src/niri/target/x86_64-unknown-linux-gnu/aur/deps/niri-af62f87027fe0571" "-Wl,--gc-sections" "-pie" "-Wl,-z,relro,-z,now" "-Wl,-O1" "-nodefaultlibs"
  = note: /usr/bin/ld: /home/user/.local/var/lib/makepkg/aur/niri-git/src/niri/target/x86_64-unknown-linux-gnu/aur/deps/niri-af62f87027fe0571.niri.1be11338d174d065-cgu.0.rcgu.o: in function `niri::pw_utils::PipeWire::start_cast::{{closure}}':
          niri.1be11338d174d065-cgu.0:(.text._ZN4niri8pw_utils8PipeWire10start_cast28_$u7b$$u7b$closure$u7d$$u7d$17h14f14741253c792aE+0x3c): undefined reference to `libspa_rs_format_parse'
          /usr/bin/ld: /home/user/.local/var/lib/makepkg/aur/niri-git/src/niri/target/x86_64-unknown-linux-gnu/aur/deps/niri-af62f87027fe0571.niri.1be11338d174d065-cgu.0.rcgu.o: in function `libspa::param::video::raw::VideoInfoRaw::parse':
          /home/user/.cargo/registry/src/index.crates.io-6f17d22bba15001f/libspa-0.7.2/src/param/video/raw.rs:375:(.text._ZN4niri8pw_utils8PipeWire10start_cast28_$u7b$$u7b$closure$u7d$$u7d$17h14f14741253c792aE+0xb3): undefined reference to `libspa_rs_format_video_raw_parse'
          collect2: error: ld returned 1 exit status

  = note: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
  = note: use the `-l` flag to specify native libraries to link
  = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorustc-link-libkindname)

error: could not compile `niri` (bin "niri") due to previous error
==> ERROR: A failure occurred in build().
    Aborting...
```

I have little to no idea what exactly is happening there. I'd guess some missing packages.

**Comment by YaLTeR** [comment]
Uh, I don't think niri changed any pipewire deps, maybe they got split in Arch in some way? Check pipewire and libspa development packages.

**Reply by muchgitsuchwow** [reply]
Hm. No idea. I don't find anything in Arch related to libspa. Arch mostly does not split things. Last pipewire installation was on  Sat 03 Feb 2024, my last successfull build of niri was 4 days later (Wed 07 Feb 2024).

In the end I'm glad it's about pipewire and not gnome, or something. By the way, don't get me wrong, (and as always keep in mind that I really have little idea about all these wayland things).. why gnome, and not wlroots? I'd love to use portal-wlr over gnome, which (on Arch) requires to install gnome, which.. I have to install a full-blown DE to run a minimalistic window manager only for being able to screenshare. I guess portal-wlr will probably never be an option, right?

(Sorry to mis-use this discussion.)

For the build issue I'll try to dig deeper into pipewire/libspa things with Arch. Thanks for the very quick response!

**Reply by muchgitsuchwow** [reply]
libpipewire has a file `libspa-0.2.pc` which seems to configure pkgconf to look in `/usr/include/spa-0.2/`, which contains some development files in a `spa/` sub directory. Hmhmhm.

**Reply by YaLTeR** [reply]
> why gnome, and not wlroots?

I like the xdp-gnome dialog, and it works well. I've been hearing that the -wlr portal screencasting does not in fact work well. Plus, well, I'm a GNOME developer, I know my way around.

> which (on Arch) requires to install gnome

Not really? Just the required libraries. If you mean gnome-desktop-4, well, that's not GNOME, that's also just a small library.

**Reply by muchgitsuchwow** [reply]
These are some very strong reasons indeed :) -- So I was just curious. I'm used to sway with portal-wlr working very well for me in the past. Thing is that niri fits my workflow so much better than sway (even now in this early state of niri), so I guess I've to overcome my irrational "fear" of installing "gnome". Looking very much forward to being able to disable outputs/screens on demand. But most importantly it would be great if I could compile niri again :/

Anyhoo. Thanks a lot for the software, the chat, the quick responses :)

**Reply by YaLTeR** [reply]
I just pushed an update to pipewire-rs 0.8.0. Maybe it'll help? Idk though.

> Looking very much forward to being able to disable outputs/screens on demand.

Wdym? You should be able to change the config to add `off` into the corresponding `output "X" {` section to disable that screen.

**Reply by muchgitsuchwow** [reply]
I'd need the screen thing on demand. In sway I have this:

```
bindswitch --reload --locked lid:on output eDP-1 disable
bindswitch --reload --locked lid:off output eDP-1 enable
```

Thanks for the hint that "hard-coding" this in the config file is possible.

With libspa-0.8.0 I've a slightly different error when compiling niri. The error looks quite identical, except that there is now an `undefined reference to spa_format_video_raw_parse_libspa_rs`. `/usr/include/spa-0.2/spa/param/video/raw-utils.h` contains `spa_format_video_raw_parse(const struct spa_pod *format,`. To me it looks like that everything is in (probably a wrong) place.

**Reply by YaLTeR** [reply]
Oh yeah, disabling internal display on lid close is something that would be good to implement to happen automatically. Smithay recently got the lid switch events which should make that possible.

**Reply by nebulosa2007** [reply]
Read with interest this thread, but not understood which errors of building are you talking about. Everyday I building niri with fresh commits (I'm this AUR package [niri-git](https://0x0.st/Hk7g.txt)) and have no building errors at all.

Btw, I'm using Niri on a laptop and on a lid close it suspend the system with blank screen. Maybe it's not because of the Niri, but it's already working as it should, I guess..

**Reply by YaLTeR** [reply]
Nah, this is about disabling the internal display when you have an external monitor connected and close the lid.

**Reply by muchgitsuchwow** [reply]
Exactly. I've disabled lid close events with systemd-logind `HandleLidSwitch=ignore`. When on my notebook (with sway) I can disable the output when closing the lid.

**Comment by muchgitsuchwow** [comment]
Funny. The PKGBUILD calls `cargo build --profile aur --offline --target x86_64-unknown-linux-gnu` and fails. Running the same command in the src directory from my shell works. Aaaand I did update my makepkg configuration according to the package defaults one day after my last successfull build. (On a side note: Such updates happen like once every 5 years or so...)

I'd say these are all strong hints that my issue here is complete Arch/AUR/makepkg- related. IF I find the root cause, I'll drop it here and close the discussion.

**Reply by nebulosa2007** [reply]
You are using obsolete PKGBUILD. Please use this [PKGBUILD](https://0x0.st/Hk7g.txt)

**Reply by muchgitsuchwow** [reply]
Ah, you added the assets :) 

I was looking for an update, not sure when this was exactly. Thanks. I'll update the PKGBUILD as well.

**Reply by nebulosa2007** [reply]
I'm maintain [niri](https://aur.archlinux.org/packages/niri) and [niri-bin](https://aur.archlinux.org/packages/niri-bin), but not [niri-git](https://aur.archlinux.org/packages/niri-git). Since september (last commit on niri-git) many changes happened, I commented it already [here](https://aur.archlinux.org/packages/niri-git#comment-955220) but nothing changed yet.

**Reply by muchgitsuchwow** [reply]
That explains a lot. Thanks for clarification. The linked PKGBUILD did work, both with `lto` and `!lto` (which I thought might make a difference). 

That's the reason why I did modify the PKGBUILD myself to integrate the assets into the package. Any chance you can adopt niri-git PKGBUILD? That'll be great. Anyhoo: Thanks both of you.

**Reply by nebulosa2007** [reply]
No, `lto` not an issue, because `rustc` has own LTO options. The reason that I added option `!lto` is Arch's [devtools](https://aur.archlinux.org/packages/niri#comment-953699) only.

> Any chance you can adopt niri-git PKGBUILD? That'll be great.

I think the best option would be to contact the niri-git maintainer directly or wait some more time... I'm still considering
